dnsprovider/aws_test.go

```diff
@@ -0,0 +1,700 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dnsprovider
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+// Compile time check for interface conformance
+var _ Route53API = &Route53APIStub{}
+
+// Route53APIStub is a minimal implementation of Route53API, used primarily for unit testing.
+// See http://http://docs.aws.amazon.com/sdk-for-go/api/service/route53.html for descriptions
+// of all of its methods.
+// mostly taken from: https://github.com/kubernetes/kubernetes/blob/853167624edb6bc0cfdcdfb88e746e178f5db36c/federation/pkg/dnsprovider/providers/aws/route53/stubs/route53api.go
+type Route53APIStub struct {
+	zones      map[string]*route53.HostedZone
+	recordSets map[string]map[string][]*route53.ResourceRecordSet
+}
+
+// NewRoute53APIStub returns an initialized Route53APIStub
+func NewRoute53APIStub() *Route53APIStub {
+	return &Route53APIStub{
+		zones:      make(map[string]*route53.HostedZone),
+		recordSets: make(map[string]map[string][]*route53.ResourceRecordSet),
+	}
+}
+
+func (r *Route53APIStub) ListResourceRecordSetsPages(input *route53.ListResourceRecordSetsInput, fn func(p *route53.ListResourceRecordSetsOutput, lastPage bool) (shouldContinue bool)) error {
+	output := route53.ListResourceRecordSetsOutput{} // TODO: Support optional input args.
+	if len(r.recordSets) <= 0 {
+		output.ResourceRecordSets = []*route53.ResourceRecordSet{}
+	} else if _, ok := r.recordSets[*input.HostedZoneId]; !ok {
+		output.ResourceRecordSets = []*route53.ResourceRecordSet{}
+	} else {
+		for _, rrsets := range r.recordSets[*input.HostedZoneId] {
+			for _, rrset := range rrsets {
+				output.ResourceRecordSets = append(output.ResourceRecordSets, rrset)
+			}
+		}
+	}
+	lastPage := true
+	fn(&output, lastPage)
+	return nil
+}
+
+func (r *Route53APIStub) ChangeResourceRecordSets(input *route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error) {
+	output := &route53.ChangeResourceRecordSetsOutput{}
+	recordSets, ok := r.recordSets[*input.HostedZoneId]
+	if !ok {
+		recordSets = make(map[string][]*route53.ResourceRecordSet)
```

Comment:
good point. However, this check is only for the internal storage of the mock. Zones are stored in the other member. I added a check so that it fails when the zone doesn't exist.